### PR TITLE
Add rewards module

### PR DESCRIPTION
**Pull request by jordy25519**
Adds a new module responsible for staker reward payouts

### CODE REVIEW

File: crml/rewards/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+decl_storage! {
+	trait Store for Module<T: Trait> as Rewards {
+		/// Inflation rate % to apply on reward payouts, it may be negative
+		pub InflationRate get(fn inflation_rate): FixedI128 = FixedI128::saturating_from_integer(1);
+		/// Development fund % take for reward payouts, parts-per-billion
+		pub DevelopmentFundTake get(fn development_fund_take) config(): Perbill;
+		/// Accumulated transaction fees for reward payout
+		pub TransactionFeePot get(fn transaction_fee_pot): BalanceOf<T>;
+		/// Historic accumulated transaction fees on reward payout
+		pub TransactionFeePotHistory get(fn transaction_fee_pot_history): VecDeque<BalanceOf<T>>;
```
Comment (aliXsed) [inline crml/rewards/src/lib.rs:77]: Is this going to grow infinitely?

Comment (jordy25519) [reply crml/rewards/src/lib.rs:77]: no, only the most recent will be kept in state

File: crml/rewards/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+		// This ensures only one storage update to total issuance will happen when dropped.
+		let mut total_payout_imbalance = <PositiveImbalanceOf<T>>::zero();
+
+		validator_commission_stake_map
+			.iter()
+			.flat_map(|(validator, validator_commission, stake_map)| {
+				Self::calculate_npos_payouts(&validator, *validator_commission, stake_map, total_payout_share)
+			})
+			.for_each(|(account, payout)| {
+				total_payout_imbalance.maybe_subsume(T::CurrencyToReward::deposit_into_existing(&account, payout).ok());
```
Comment (aliXsed) [inline crml/rewards/src/lib.rs:151]: I suppose when a fee is paid for an extrinsic, we burn that and that's why we need to create an imbalance for the whole reward payout. Is that true?

File: crml/rewards/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+use frame_system::{self as system, ensure_root};
+use sp_arithmetic::{FixedI128, FixedPointNumber, FixedPointOperand};
+use sp_runtime::{
+	traits::{AccountIdConversion, One, Saturating, Zero},
+	ModuleId, Perbill,
+};
+use sp_std::{collections::vec_deque::VecDeque, prelude::*};
+
+/// The number of historical eras for which tx fee payout info should be retained.
+const HISTORICAL_PAYOUT_ERAS: usize = 7;
```
Comment (aliXsed) [inline crml/rewards/src/lib.rs]: How about having a runtime parameter for this? -> parameter_types!

File: crml/rewards/src/lib.rs (hunk 2 above)
Comment (jordy25519) [reply crml/rewards/src/lib.rs:151]: yes, it will reissue those fee tokens

File: crml/rewards/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+		// Deduct development fund take %
+		let development_fund_payout = Self::development_fund_take() * total_payout;
+		let _ = T::CurrencyToReward::deposit_into_existing(
+			&T::TreasuryModuleId::get().into_account(),
+			development_fund_payout,
+		);
+		total_payout = total_payout.saturating_sub(development_fund_payout);
+
+		// Payout reward to validators and their nominators
+		let total_payout_share = total_payout / BalanceOf::<T>::from(validator_commission_stake_map.len() as u32);
```
Comment (aliXsed) [inline crml/rewards/src/lib.rs:138]: This https://github.com/cennznet/cennznet/pull/304/files#diff-8b6330393ec7c1d63275edb851ba0d647bc896357706fa9377fe211ae033f0abR120-R139 
could be an individual function for calculating the payout share to each validator and its nominators.

File: crml/rewards/src/lib.rs (hunk 2 above)
Comment (aliXsed) [inline crml/rewards/src/lib.rs:151]: I guess it could be better that another individual function of StakerRewardPayment just returns "who should get how much" and then the actual deposit and then event generation is left to the Module. The more we can separate concerns, the better we can change our strategies in the future.

File: crml/rewards/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+				Rewards::inflation_rate(),
+				FixedI128::saturating_from_rational(-1, 1_000)
+			)
+		});
+	}
+
+	#[test]
+	fn set_inflation_rate_bounds() {
+		ExtBuilder::default().build().execute_with(|| {
+			assert_err!(
```
Comment (aliXsed) [inline crml/rewards/src/lib.rs]: assert_noop might be better if we don't expect any change in the storage.

File: crml/rewards/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,727 @@
...
+			assert_ok!(Rewards::set_inflation_rate(
+				Origin::root(),
+				i64::max_value(),
+				i64::max_value()
+			));
+			assert_ok!(Rewards::set_inflation_rate(
+				Origin::root(),
+				i64::min_value(),
+				i64::min_value()
+			));
```
Comment (aliXsed) [inline crml/rewards/src/lib.rs:460]: Is it worth testing the mix of min and max?

File: primitives/src/traits.rs [hunk 7]
```diff
@@ -69,3 +76,20 @@ pub trait SimpleAssetSystem {
 	/// Get the default asset/currency ID in the system
 	fn default_asset_id() -> Self::AssetId;
 }
+
+/// Something which can perform reward payment to staked validators
+pub trait StakerRewardPayment {
```
Comment (aliXsed) [inline primitives/src/traits.rs:81]: There is a potential for "primitives" to become a place for holding random unrelated stuff.

File: primitives/src/types.rs [hunk 8]
```diff
@@ -116,3 +117,28 @@ impl<AssetId: Copy, Balance: Copy> FeeExchange<AssetId, Balance> {
...
+	/// The stash account of the nominator in question.
+	pub who: AccountId,
+	/// Amount of funds exposed.
+	#[codec(compact)]
+	pub value: Balance,
+}
+
+/// A snapshot of the stake backing a single validator in the system.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Decode, Default, RuntimeDebug)]
+pub struct Exposure<AccountId, Balance: HasCompact> {
```
Comment (aliXsed) [inline primitives/src/types.rs:135]: The same here, the types that could be better located in the Staking pallet now moved to be next to other unrelated stuff. I suppose you need to put them here so you share them between rewards and staking pallets. However, I'm asking this question if rewards could be a separate unit inside the staking module, the same as reward_curve?

Comment (jordy25519) [reply primitives/src/types.rs:135]: The calculations could be, the main reason for separating rewards and staking like this to isolate the reward related storage and config from the staking storage and config, it made sense to bring the calculations also.
\+ the staking module is on the big side already

File: primitives/src/traits.rs (hunk 7 above)
Comment (jordy25519) [reply primitives/src/traits.rs:81]: that's what it's used for, we could rename it to something besides `primitives` to indicate that better?

File: primitives/src/types.rs (hunk 8 above)
Comment (aliXsed) [reply primitives/src/types.rs:135]: We could even separate the storage and config in the same way that we have done for Sylo within the same pallet. At the end of the day when a service like staking is complex, it's true that we should decompose it to smaller components but we don't necessarily need to then scatter those components from the only location they can have a meaningful place. If there was a different pallet or service that could run with rewards but without the staking module, it could make sense. But now I'm not sure.

**Comment by aliXsed** [comment]
I have left a few comments regarding how best we can structure this change. As those comments are just suggestions, I'm going to approve this PR at this stage. I'm hoping the next PR for integrating this module with the rest of cennznet can show that there haven't been any behavioural changes as otherwise, I couldn't easily note and evaluate them through looking at this change by itslef.

**Review by aliXsed — APPROVED** [review]

File: primitives/src/types.rs (hunk 8 above)
Comment (jordy25519) [reply primitives/src/types.rs:135]: Is this suggestion about where the rust files/modules sit? e.g.
```
# now
crml/rewards
crml/staking
# after
crml/staking/
crml/staking/rewards/<reward crate>
```
if so then yes I should be able to make this change

**Comment by jordy25519** [comment]
There are a few rough edges on the rewards API, it's not so much generic as it is decoupled. So there's definitely room to improve it, including the individual reward calculation could come in handy

File: primitives/src/types.rs (hunk 8 above)
Comment (aliXsed) [reply primitives/src/types.rs:135]: Yup correct
